src/CodeGen_D3D12Compute_Dev.cpp

```diff
@@ -0,0 +1,1154 @@
+            << " = " << simt_intrinsic(loop->name) << ";\n";
+
+    loop->body.accept(this);
+}
+
+void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::visit(const Ramp *op) {
+    ostringstream rhs;
+    rhs << print_expr(op->base)
+        << " + "
+        << print_expr(op->stride)
+        << " * "
+        << print_type(op->type.with_lanes(op->lanes))
+        << "(";
+
+    rhs << "0";
+    for (int i = 1; i < op->lanes; ++i) {
+        rhs << ", " << i;
+    }
+
+    rhs << ")";
+    print_assignment(op->type.with_lanes(op->lanes), rhs.str());
+}
+
+void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::visit(const Broadcast* op)
+{
+    string id_value = print_expr(op->value);
+    user_assert(op->value.type().lanes() == 1) << "Broadcast source must be 1-wide.\n";
+
+    ostringstream rhs;
+    rhs << print_type(op->type.with_lanes(op->lanes))
+        << "(";
+    for (int i = 0; i < op->lanes; ++i)
+    {
+        rhs << id_value;
+        if (i < op->lanes-1)
+            rhs << ", ";
+    }
+    rhs << ")";
+
+    print_assignment(op->type.with_lanes(op->lanes), rhs.str());
+}
+
+namespace {
+
+// If e is a ramp expression with stride 1, return the base, otherwise undefined.
+Expr is_ramp_one(Expr e) {
+    const Ramp *r = e.as<Ramp>();
+    if (r == nullptr) {
+        return Expr();
+    }
+
+    if (is_one(r->stride)) {
+        return r->base;
+    }
+
+    return Expr();
+}
+
+template<typename T>
+string hex_literal(T value)
+{
+    ostringstream hex;
+    hex << "0x" << std::uppercase << std::setfill('0') << std::setw(8) << std::hex
+        << value;
+    return(hex.str());
+}
+
+}
+
+void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::visit(const Load* op)
+{
+    user_assert(is_one(op->predicate)) << "Predicated load is not supported inside D3D12Compute kernel.\n";
+
+    // __shared[x] is always uint(32): must reinterpret/unpack bits...
+    if (op->name == "__shared")
+    {
+        ostringstream rhs;
+        internal_assert(allocations.contains(op->name));
+        // no ramps when accessing shared memory...
+        Expr ramp_base = is_ramp_one(op->index);
+        internal_assert(!ramp_base.defined());
+        // shared memory in Halide is represented as a byte buffer
+        // but the 'op->index' is actually in terms of elements...
+        // to complicate things, HLSL (SM 5.1) only supports 32bit
+        // words (int/uint/float) as groupshared types...
+        internal_assert(allocations.get(op->name).type == UInt(8));
+        internal_assert(op->type.lanes() == 1);
+#if 1
+        string id_index = print_expr(op->index);
+        //id_index = print_cast(op->type, op->index.type(), id_index);
+        internal_assert(op->type.bits() <= 32);
+        Type promoted = op->type.with_bits(32);
+        rhs << "as" << print_type(promoted)
+            << "("
+            << print_name(op->name)
+            << "[" << id_index << "]"
+            << ")";
+#else
```

Comment:
Possibly accidentally left this in?  If not, please add comment explaining why this is here